src/packer.ts

```diff
@@ -0,0 +1,195 @@
+import array from "@stdlib/ndarray/array";
+import ndarray2array from "@stdlib/ndarray/to-array";
+import MultiSlice from "@stdlib/slice/multi";
+import Slice from "@stdlib/slice/ctor";
+import slice from "@stdlib/ndarray/slice";
+import {ndarray} from "@stdlib/types/ndarray";
+
+export interface PackerOptions {
+    // names of scalar values
+    scalar: Set<string>,
+
+    // mapping of array value names to their length. Arrays may be multi-dimensional so the value of the map may be an
+    // array containing lengths of each dimension. 0 or null values are counted as scalars, which allows you to put
+    // scalars at positions other than the front of the packing vector.
+    array: Map<string, number | number[] | null>,
+
+    // mapping of fixed data names to values
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    fixed: Map<string, any>,
+
+    // custom function to do additional unpacking logic on the unpacked map
+    // TODO: sort out types..
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    process: (unpacked: Map<string, any>) => Map<string, any>
+}
+
+// For a given name and associated shape, do some validation, make shape type consistent and return number of values
+// after expansion
+const preparePackArrayForShape = (name: string, shape: number |  number[] | null) => {
+    // Scalars can be defined within the array option, if that allows user's preferred ordering
+    const scalar = shape === null || (Array.isArray(shape) && shape.length === 0);
+    if (scalar) {
+        return { names: [name], shape: [0], n: 1 };
+    }
+    const arrayShape: Array<number> = typeof shape === "number" ? [shape] : shape;
+    const nonInteger = arrayShape.find((v) => !Number.isInteger(v))
+    if (nonInteger) {
+        throw Error("All dimension values in 'array' values must be integers, but this is not the case for " +
+                    `${name}, whose value is ${JSON.stringify(shape)}`);
+    }
+    const lessThanZero = arrayShape.find((v) => v <= 0);
+    if (lessThanZero) {
+        throw Error("All dimension values in 'array' values must be at least 1, but this is not the case for " +
+                    `${name}, whose value is ${JSON.stringify(shape)}`);
+    }
+
+    // total number of values in shape
+    const n = arrayShape.reduce((prev, current) => prev * current, 1);
+
+    return {
+        shape: arrayShape,
```

Comment:
this does not return `names: [name]` like the early exit does - is that intended?